### PR TITLE
Add multi-class, multi-label and regression to transformers

### PR DESCRIPTION
**Pull request by abhishekkrthakur**
This PR adds support for single/multi column regression and single/multi label classification tasks to `SequenceClassification` models. The `problem_type` can be specified in the config: `regression`, `single_label_classification`, `multi_label_classification`.

### CODE REVIEW

File: src/transformers/models/bert/modeling_bert.py [hunk 1]
```diff
@@ -1517,13 +1518,29 @@ def forward(
...
-                #  We are doing regression
-                loss_fct = MSELoss()
-                loss = loss_fct(logits.view(-1), labels.view(-1))
+            if self.problem_type is not None:
+                if self.problem_type == "single_column_regression":
+                    loss_fct = MSELoss()
+                    loss = loss_fct(logits.view(-1), labels.view(-1))
+                elif self.problem_type == "multi_column_regression":
+                    loss_fct = MSELoss()
+                    loss = loss_fct(logits.view(-1, self.num_labels), labels)
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_bert.py:1527]: I would group those into just `regression`. `loss = loss_fct(logits.view(-1, self.num_labels), labels)` will work in both cases and we can find out if really need to if there is one or several labels by looking at `config.num_labels.`

File: src/transformers/models/bert/modeling_bert.py [hunk 2]
```diff
@@ -1517,13 +1518,29 @@ def forward(
...
+                elif self.problem_type in ("binary_classification", "multi_class_classification"):
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_bert.py:1528]: I would group those in `single_label_classification`, no need for two names, especially if we can find by looking at the number of labels if we really are in a binary classification setting.

File: src/transformers/models/bert/modeling_bert.py [hunk 3]
```diff
@@ -1517,13 +1518,29 @@ def forward(
...
             else:
-                loss_fct = CrossEntropyLoss()
-                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
+                if self.num_labels == 1:
+                    #  We are doing regression
+                    loss_fct = MSELoss()
+                    loss = loss_fct(logits.view(-1), labels.view(-1))
+                else:
+                    loss_fct = CrossEntropyLoss()
+                    loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_bert.py:1543]: I was thinking that we should set `problem_type` to a default value if it's None:
- one label or float labels -> regression
- int labels of shape [seq_len] -> single_label_classification
- array of labels of shape [seq_len x num_labels] -> multi-label classification

This way we still save in the config how the model was trained.

**Review by sgugger — APPROVED** [review]
Looking good! I would juste create an enum for the `ProblemType` values and simplify them to just three values instead of five.

File: src/transformers/models/bert/modeling_bert.py (hunk 3 above)
Comment (abhishekkrthakur) [reply src/transformers/models/bert/modeling_bert.py:1543]: > array of labels of shape [seq_len x num_labels] -> multi-label classification
this can also mean multi-column regression. i can check dtype though: float => regression, else classification. WDYT?

Comment (sgugger) [reply src/transformers/models/bert/modeling_bert.py:1543]: Doesn't work as BCEWithLogitsLoss wants the targets as floats. So we have to pick one default between multi-label and regression for that case. I'd go with multi-label personally.

File: src/transformers/models/bert/modeling_bert.py [hunk 4]
```diff
@@ -1517,13 +1518,29 @@ def forward(
...
+                    loss_fct = MSELoss()
+                    loss = loss_fct(logits.view(-1, self.num_labels), labels)
+                elif self.problem_type in ("binary_classification", "multi_class_classification"):
+                    loss_fct = CrossEntropyLoss()
+                    loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
+                elif self.problem_type in ("multi_label_classification"):
+                    loss_fct = BCEWithLogitsLoss()
+                    loss = loss_fct(logits, labels)
+                else:
+                    raise Exception("Problem type not understood")
```
Comment (LysandreJik) [inline src/transformers/models/bert/modeling_bert.py:1535]: This should throw a more specific exception like `ValueError`

Comment (LysandreJik) [reply src/transformers/models/bert/modeling_bert.py:1535]: I'd be a bit more explicit on the error text too:
```suggestion
                    raise ValueError(f"The config parameter `problem_type` not understood: received {self.problem_type} but only [xxx, xxx] are valid.") 
```
(xxx, xxx to be replaced, of course)

**Review by LysandreJik — APPROVED** [review]
Yes, after discussing it  w/ @sgugger & thinking about it I think this is a good way to tackle the problem. In the end, we would have `config.problem_type` work as a similar parameter as `config.num_labels`, which is fine for me.

I also like that it's task agnostic (`problem_type` vs `classification_type`) as it offers compatibility with other heads if the need ever arises.

I agree with Sylvain on his comments, my comments are only nits as you plan on applying this change to other models.

File: src/transformers/models/bert/modeling_bert.py (hunk 4 above)
Comment (abhishekkrthakur) [reply src/transformers/models/bert/modeling_bert.py]: fixed

File: src/transformers/models/bert/modeling_bert.py (hunk 3 above)
Comment (abhishekkrthakur) [reply src/transformers/models/bert/modeling_bert.py]: fixed

File: src/transformers/models/bert/modeling_bert.py (hunk 2 above)
Comment (abhishekkrthakur) [reply src/transformers/models/bert/modeling_bert.py]: fixed

File: src/transformers/models/bert/modeling_bert.py (hunk 1 above)
Comment (abhishekkrthakur) [reply src/transformers/models/bert/modeling_bert.py]: fixed

File: src/transformers/configuration_utils.py [hunk 5]
```diff
@@ -152,6 +152,8 @@ class PretrainedConfig(object):
           typically for a classification task.
         - **task_specific_params** (:obj:`Dict[str, Any]`, `optional`) -- Additional keyword arguments to store for the
           current task.
+        - **problem_type** (:obj:`str`, `optional`) -- Problem type for ForSequenceClassification tasks. It can be one
+          of (None, "regression", "single_label_classification", "multi_label_classification"). Default is None.
```
Comment (sgugger) [inline src/transformers/configuration_utils.py:156]:
```suggestion
        - **problem_type** (:obj:`str`, `optional`) -- Problem type for :obj:`XxxForSequenceClassification` models. Can be one
          of (:obj:`"regression"`, :obj:`"single_label_classification"`, :obj:`"multi_label_classification"`).
```
For proper formatting of the strings compared to the rest of the doc. Also no need to say it defaults to None, it's implied if no default is specified (we do this consistently in the rest of the documentation).

File: src/transformers/configuration_utils.py [hunk 6]
```diff
@@ -249,6 +251,16 @@ def __init__(self, **kwargs):
         # task specific arguments
         self.task_specific_params = kwargs.pop("task_specific_params", None)
 
+        # regression / multi-label classification
+        self.problem_type = kwargs.pop("problem_type", None)
+        allowed_problem_types = (None, "regression", "single_label_classification", "multi_label_classification")
+        if self.problem_type not in allowed_problem_types:
```
Comment (sgugger) [inline src/transformers/configuration_utils.py:257]:
```suggestion
        allowed_problem_types = ("regression", "single_label_classification", "multi_label_classification")
        if self.problem_type is not None and self.problem_type not in allowed_problem_types:
```
Absolute nit, but I would keep the None separate.

File: src/transformers/models/bert/modeling_bert.py [hunk 7]
```diff
@@ -1517,14 +1518,29 @@ def forward(
...
         if labels is not None:
-            if self.num_labels == 1:
-                #  We are doing regression
+            if self.problem_type is None:
+                if self.num_labels == 1:
+                    self.problem_type = "regression"
+                elif self.num_labels > 1 and type(labels) == torch.long:
+                    self.problem_type = "single_label_classification"
+                else:
+                    self.problem_type = "multi_label_classification"
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_bert.py:1527]: The problem here is that it will be set as attribute to the model but not config attribute, so when saving the config, it won't be saved :-(

I think the easiest is to remove the shortcut `self.problem_type` and always use `self.config.problem_type`.

**Review by sgugger — COMMENTED** [review]
Thanks for the changes, this is all in the right direction in my opinion! Left a few more comments.

File: src/transformers/models/bert/modeling_bert.py (hunk 7 above)
Comment (abhishekkrthakur) [reply src/transformers/models/bert/modeling_bert.py:1527]: done

File: src/transformers/models/albert/modeling_albert.py [hunk 8]
```diff
@@ -1024,13 +1025,27 @@ def forward(
...
+            elif self.config.problem_type == "single_label_classification":
                 loss_fct = CrossEntropyLoss()
                 loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
+            elif self.config.problem_type == "multi_label_classification":
+                loss_fct = BCEWithLogitsLoss()
+                loss = loss_fct(logits, labels)
+            else:
+                raise ValueError(
+                    f"The config parameter `problem_type` not understood: received {self.config.problem_type} but only [regression, single_label_classification and multi_label_classification] are valid."
+                )
```
Comment (sgugger) [inline src/transformers/models/albert/modeling_albert.py:1048]: This has already been raised in the config, not sure it needs to be re-raised here.
In any case, the text is too long so needs to be split in two or three lines (119 char limits) and have the same corrections as in the config.

Change then needs to be applied to the other modeling files.

File: src/transformers/configuration_utils.py [hunk 9]
```diff
@@ -260,6 +270,18 @@ def __init__(self, **kwargs):
...
+        self.problem_type = kwargs.pop("problem_type", None)
+        allowed_problem_types = ("regression", "single_label_classification", "multi_label_classification")
+        if self.problem_type is not None and self.problem_type not in allowed_problem_types:
+            raise ValueError(
+                f"""
+The config parameter `problem_type` not understood:
+                    received {self.problem_type} but only [regression, single_label_classification and
+                    multi_label_classification] are valid.
+                    """
+            )
```
Comment (sgugger) [inline src/transformers/configuration_utils.py:283]:
```suggestion
            raise ValueError(
                f"The config parameter `problem_type` wasnot understood: received {self.problem_type}"
                "but only 'regression', 'single_label_classification' and 'multi_label_classification' are valid."
            )
```

File: src/transformers/configuration_utils.py [hunk 10]
```diff
@@ -163,6 +163,16 @@ class PretrainedConfig(PushToHubMixin):
...
+        - **problem_type** (:obj:`str`, `optional`) -- Problem type for :obj:`XxxForSequenceClassification` models. Can
+          be one of (:obj:`"regression"`, :obj:`"single_label_classification"`, :obj:`"multi_label_classification"`).
+          Please note that this parameter is only available in the following models:
+          `"AlbertForSequenceClassification"`, `"BertForSequenceClassification"`, `"BigBirdForSequenceClassification"`,
+          `"ConvBertForSequenceClassification"`, `"DistilBertForSequenceClassification"`,
+          `"ElectraForSequenceClassification"`, `"FunnelForSequenceClassification"`,
+          `"LongformerForSequenceClassification"`, `"MobileBertForSequenceClassification"`,
+          `"ReformerForSequenceClassification"`, `"RobertaForSequenceClassification"`,
+          `"SqueezeBertForSequenceClassification"`, `"XLMForSequenceClassification"` and
+          `"XLNetForSequenceClassification"`.
```
Comment (sgugger) [inline src/transformers/configuration_utils.py:175]: The `"xxx"` will render in italics, not code-formatting.

**Review by sgugger — APPROVED** [review]
LGTM!
